File: src/core/zig-manifest.ts
import { dependencyEntry } from './dependency';
import type { DependencyEntry, SourceRange } from './model';

type TokenKind = 'identifier' | 'string' | 'symbol';

interface Token {
  kind: TokenKind;
  value: string;
  start: number;
  end: number;
  contentStart?: number;
  contentEnd?: number;
}

interface ZonField {
  name: string;
  primaryRange: SourceRange;
  sourceRange: SourceRange;
  iconRange: SourceRange;
  equalsIndex: number;
}

const OPEN_TO_CLOSE: Readonly<Record<string, string>> = {
  '{': '}',
  '[': ']',
  '(': ')',
};

function decodeString(value: string): string {
  let out = '';
  for (let cursor = 0; cursor < value.length; cursor++) {
    const char = value[cursor]!;
    if (char !== '\\' || cursor + 1 >= value.length) {
      out += char;
      continue;
    }
    const escaped = value[++cursor]!;
    if (escaped === 'n') out += '\n';
    else if (escaped === 'r') out += '\r';
    else if (escaped === 't') out += '\t';
    else if (escaped === 'x') {
      const digits = value.slice(cursor + 1, cursor + 3);
      const code = /^[0-9A-Fa-f]{2}$/.test(digits) ? Number.parseInt(digits, 16) : undefined;
      if (code === undefined) out += 'x';
      else {
        out += String.fromCodePoint(code);
        cursor += 2;
      }
    } else if (escaped === 'u' && value[cursor + 1] === '{') {
      const close = value.indexOf('}', cursor + 2);
      const digits = close < 0 ? '' : value.slice(cursor + 2, close);
      const code = /^[0-9A-Fa-f]{1,6}$/.test(digits) ? Number.parseInt(digits, 16) : undefined;
      if (code === undefined || code > 0x10ffff) out += 'u';
      else {
        out += String.fromCodePoint(code);
        cursor = close;
      }
    } else {
      out += escaped;
    }
  }
  return out;
}

/** A small, tolerant ZON lexer. Pacmon reads manifests as text and never invokes Zig. */
function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let cursor = 0;
  while (cursor < text.length) {
    const char = text[cursor]!;
    if (/\s/.test(char)) {
      cursor++;
      continue;
    }
    if (char === '/' && text[cursor + 1] === '/') {
      cursor += 2;
      while (cursor < text.length && text[cursor] !== '\r' && text[cursor] !== '\n') cursor++;
      continue;
    }
    // A Zig multiline-string line starts with two backslashes and runs to EOL.
    if (char === '\\' && text[cursor + 1] === '\\') {
      cursor += 2;
      while (cursor < text.length && text[cursor] !== '\r' && text[cursor] !== '\n') cursor++;
      continue;
    }
    if (char === '"' || char === "'") {
      const quote = char;
      const start = cursor++;
      const contentStart = cursor;
      while (cursor < text.length) {
        if (text[cursor] === '\\' && cursor + 1 < text.length) {
          cursor += 2;
          continue;
        }
        if (text[cursor] === quote) break;
        cursor++;
      }
      const contentEnd = cursor;
      if (cursor < text.length) cursor++;
      const raw = text.slice(contentStart, contentEnd);
      tokens.push({
        kind: 'string',
        value: quote === '"' ? decodeString(raw) : raw,
        start,
        end: cursor,
        contentStart,
        contentEnd,
      });
      continue;
    }
    if (/[A-Za-z_]/.test(char)) {
      const start = cursor++;
      while (cursor < text.length && /[A-Za-z0-9_]/.test(text[cursor]!)) cursor++;
      tokens.push({ kind: 'identifier', value: text.slice(start, cursor), start, end: cursor });
      continue;
    }
    tokens.push({ kind: 'symbol', value: char, start: cursor, end: cursor + 1 });
    cursor++;
  }
  return tokens;
}

function fieldAt(tokens: readonly Token[], dotIndex: number): ZonField | undefined {
  const dot = tokens[dotIndex];
  if (dot?.value !== '.') return undefined;
  const next = tokens[dotIndex + 1];
  if (!next) return undefined;

  let name: string;
  let primaryRange: SourceRange;
  let sourceRange: SourceRange;
  let equalsIndex: number;
  if (next.kind === 'identifier') {
    name = next.value;
    primaryRange = { offset: next.start, length: next.end - next.start };
    sourceRange = primaryRange;
    equalsIndex = dotIndex + 2;
  } else if (next.value === '@' && tokens[dotIndex + 2]?.kind === 'string') {
    const quoted = tokens[dotIndex + 2]!;
    name = quoted.value;
    primaryRange = {
      offset: quoted.contentStart ?? quoted.start,
      length: (quoted.contentEnd ?? quoted.end) - (quoted.contentStart ?? quoted.start),
    };
    sourceRange = { offset: next.start, length: quoted.end - next.start };
    equalsIndex = dotIndex + 3;
  } else {
    return undefined;
  }
  if (tokens[equalsIndex]?.value !== '=') return undefined;
  return {
    name,
    primaryRange,
    sourceRange,
    iconRange: { offset: dot.start, length: dot.end - dot.start },
    equalsIndex,
  };
}

function dependenciesOpen(tokens: readonly Token[]): number | undefined {
  const stack: string[] = [];
  for (let index = 0; index < tokens.length; index++) {
    const token = tokens[index]!;
    if (stack.length === 1) {
      const field = fieldAt(tokens, index);
      if (field?.name === 'dependencies' && tokens[field.equalsIndex + 1]?.value === '.' && tokens[field.equalsIndex + 2]?.value === '{') {
        return field.equalsIndex + 2;
      }
    }
    const closing = OPEN_TO_CLOSE[token.value];
    if (closing) stack.push(closing);
    else if (token.value === stack[stack.length - 1]) stack.pop();
  }
  return undefined;
}

/** Extracts direct build.zig.zon dependencies without evaluating build.zig or invoking Zig. */
export function extractZigDependencies(text: string): DependencyEntry[] {
  const tokens = tokenize(text);
  const open = dependenciesOpen(tokens);
  if (open === undefined) return [];

  const out: DependencyEntry[] = [];
  const seen = new Set<string>();
  const stack = ['}'];
  for (let index = open + 1; index < tokens.length; index++) {
    const token = tokens[index]!;
    if (stack.length === 1) {
      if (token.value === '}') break;
      const field = fieldAt(tokens, index);
      if (field && !seen.has(field.name)) {
        seen.add(field.name);
        out.push(dependencyEntry(
          field.name,
          'dependencies',
          field.primaryRange,
          [field.sourceRange],
          field.name,
          field.iconRange,
        ));
      }
    }
    const closing = OPEN_TO_CLOSE[token.value];
    if (closing) stack.push(closing);
    else if (token.value === stack[stack.length - 1]) stack.pop();
  }
  return out;
}

File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/ZigManifest.kt
package dev.pacmon.jetbrains.core

/** Static build.zig.zon parser. Pacmon never invokes Zig or evaluates build.zig. */
object ZigManifestAdapter : ManifestAdapter {
    override val kind = ManifestKind.ZIG
    override val fileNames = listOf("build.zig.zon")
    override val notesRelativePath = ".pacmon/zig/DEPENDENCY-NOTES.md"
    override fun normalizeNoteKey(raw: String): String = ManifestRegistry.stripName(raw)

    private enum class TokenKind { IDENTIFIER, STRING, SYMBOL }

    private data class Token(
        val kind: TokenKind,
        val value: String,
        val start: Int,
        val end: Int,
        val contentStart: Int? = null,
        val contentEnd: Int? = null,
    )

    private data class ZonField(
        val name: String,
        val primaryRange: SourceRange,
        val sourceRange: SourceRange,
        val iconRange: SourceRange,
        val equalsIndex: Int,
    )

    private val openToClose = mapOf("{" to "}", "[" to "]", "(" to ")")

    override fun extractDependencies(text: String): List<DependencyEntry> {
        val tokens = tokenize(text)
        val open = dependenciesOpen(tokens) ?: return emptyList()
        val out = mutableListOf<DependencyEntry>()
        val seen = mutableSetOf<String>()
        val stack = mutableListOf("}")

        for (index in open + 1 until tokens.size) {
            val token = tokens[index]
            if (stack.size == 1) {
                if (token.value == "}") break
                val field = fieldAt(tokens, index)
                if (field != null && seen.add(field.name)) {
                    out.add(
                        dependencyEntry(
                            noteKey = field.name,
                            scope = "dependencies",
                            primaryRange = field.primaryRange,
                            sourceRanges = listOf(field.sourceRange),
                            displayName = field.name,
                            iconRange = field.iconRange,
                        ),
                    )
                }
            }
            openToClose[token.value]?.let(stack::add)
                ?: if (token.value == stack.lastOrNull()) stack.removeAt(stack.lastIndex) else Unit
        }
        return out
    }

    private fun dependenciesOpen(tokens: List<Token>): Int? {
        val stack = mutableListOf<String>()
        for (index in tokens.indices) {
            val token = tokens[index]
            if (stack.size == 1) {
                val field = fieldAt(tokens, index)
                if (field?.name == "dependencies" &&
                    tokens.getOrNull(field.equalsIndex + 1)?.value == "." &&
                    tokens.getOrNull(field.equalsIndex + 2)?.value == "{"
                ) {
                    return field.equalsIndex + 2
                }
            }
            openToClose[token.value]?.let(stack::add)
                ?: if (token.value == stack.lastOrNull()) stack.removeAt(stack.lastIndex) else Unit
        }
        return null
    }

    private fun fieldAt(tokens: List<Token>, dotIndex: Int): ZonField? {
        val dot = tokens.getOrNull(dotIndex)?.takeIf { it.value == "." } ?: return null
        val next = tokens.getOrNull(dotIndex + 1) ?: return null
        val name: String
        val primaryRange: SourceRange
        val sourceRange: SourceRange
        val equalsIndex: Int

        if (next.kind == TokenKind.IDENTIFIER) {
            name = next.value
            primaryRange = SourceRange(next.start, next.end - next.start)
            sourceRange = primaryRange
            equalsIndex = dotIndex + 2
        } else if (next.value == "@" && tokens.getOrNull(dotIndex + 2)?.kind == TokenKind.STRING) {
            val quoted = tokens[dotIndex + 2]
            name = quoted.value
            val contentStart = quoted.contentStart ?: quoted.start
            val contentEnd = quoted.contentEnd ?: quoted.end
            primaryRange = SourceRange(contentStart, contentEnd - contentStart)
            sourceRange = SourceRange(next.start, quoted.end - next.start)
            equalsIndex = dotIndex + 3
        } else {
            return null
        }
        if (tokens.getOrNull(equalsIndex)?.value != "=") return null
        return ZonField(
            name,
            primaryRange,
            sourceRange,
            SourceRange(dot.start, dot.end - dot.start),
            equalsIndex,
        )
    }

    private fun tokenize(text: String): List<Token> {
        val tokens = mutableListOf<Token>()
        var cursor = 0
        while (cursor < text.length) {
            val char = text[cursor]
            if (char.isWhitespace()) {
                cursor++
                continue
            }
            if (char == '/' && text.getOrNull(cursor + 1) == '/') {
                cursor += 2
                while (cursor < text.length && text[cursor] != '\r' && text[cursor] != '\n') cursor++
                continue
            }
            if (char == '\\' && text.getOrNull(cursor + 1) == '\\') {
                cursor += 2
                while (cursor < text.length && text[cursor] != '\r' && text[cursor] != '\n') cursor++
                continue
            }
            if (char == '"' || char == '\'') {
                val quote = char
                val start = cursor++
                val contentStart = cursor
                while (cursor < text.length) {
                    if (text[cursor] == '\\' && cursor + 1 < text.length) {
                        cursor += 2
                        continue
                    }
                    if (text[cursor] == quote) break
                    cursor++
                }
                val contentEnd = cursor
                if (cursor < text.length) cursor++
                val raw = text.substring(contentStart, contentEnd)
                tokens.add(
                    Token(
                        TokenKind.STRING,
                        if (quote == '"') decodeString(raw) else raw,
                        start,
                        cursor,
                        contentStart,
                        contentEnd,
                    ),
                )
                continue
            }
            if (char == '_' || char.isLetter()) {
                val start = cursor++
                while (cursor < text.length &&
                    (text[cursor] == '_' || text[cursor].isLetterOrDigit())
                ) cursor++
                tokens.add(Token(TokenKind.IDENTIFIER, text.substring(start, cursor), start, cursor))
                continue
            }
            tokens.add(Token(TokenKind.SYMBOL, char.toString(), cursor, cursor + 1))
            cursor++
        }
        return tokens
    }

    private fun decodeString(value: String): String = buildString {
        var cursor = 0
        while (cursor < value.length) {
            val char = value[cursor++]
            if (char != '\\' || cursor >= value.length) {
                append(char)
                continue
            }
            when (val escaped = value[cursor++]) {
                'n' -> append('\n')
                'r' -> append('\r')
                't' -> append('\t')
                'x' -> {
                    val end = (cursor + 2).coerceAtMost(value.length)
                    val digits = value.substring(cursor, end)
                    val code = digits.takeIf { it.length == 2 && it.all { char -> char.isHexDigit() } }?.toIntOrNull(16)
                    if (code == null) append('x')
                    else {
                        append(code.toChar())
                        cursor = end
                    }
                }
                'u' -> {
                    val close = if (value.getOrNull(cursor) == '{') value.indexOf('}', cursor + 1) else -1
                    val digits = if (close < 0) "" else value.substring(cursor + 1, close)
                    val code = digits.takeIf { it.length in 1..6 && it.all { char -> char.isHexDigit() } }?.toIntOrNull(16)
                    if (code == null || code > 0x10ffff) append('u')
                    else {
                        appendCodePoint(code)
                        cursor = close + 1
                    }
                }
                else -> append(escaped)
            }
        }
    }

    private fun Char.isHexDigit(): Boolean = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
}
